enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+When a host process or host networked pod on a Kubernetes node initiates a connection to a service, iptables rules will DNAT the nodeport or loadbalancer IP into the cluster IP, and then send the traffic via br-ex where it is masqueraded and sent into the OVN GR. These flows can all be modified to detect the service IP and then send to the correct GR-&lt;network> patch port. However, in the case where there is hairpin service traffic to the host (Host->Service->Endpoint is also the host), the endpoint reply traffic will need to be distinguishable on a per network basis. In order to achieve this, each OVN GR’s unique masquerade IP will be leveraged.
+
+For service access towards KAPI/DNS or potentially other services on the default network, there are two potential technical solutions. Assume eth0 is the pod interface connected to the default network, and eth1 is connected to the unique primary network:
+
+1. Add routes for KAPI/DNS specifically into the pod to go out eth0, while all other service access will go to eth1. This will then just work normally with the load balancers on the switches for the respective networks.
+
+2. Do not send any service traffic out of eth0, instead all service traffic goes to eth1. ovn_cluster_router_&lt;unique network> is configured to send service CIDR traffic to ovn-k8s-mp0-&lt;unique network>. ipTables rules in the host only permit access to KAPI/DNS and drop all other service traffic coming from ovn-k8s-mp0-&lt;unique network>. The traffic then gets routed to br-ex and default GR where it hits the service.
+
+While the second option is more complex, it allows for not configuring routes to service addresses in the pod that could hypothetically change.
+
+
+#### Egress IP
+
+This feature works today by labeling and choosing a node+network to be egress, and then OVN logical routes and logical route policies are created which steer traffic from a pod towards a specific gateway router (for primary network egress). From there the packets are SNAT’ed by the OVN GR to the egress IP, and sent to br-ex. Egress IP is cluster scoped, but applies to selected namespaces, which will allow us to only apply the SNAT and routes to the GR and OVN topology elements of that network. In the layer 3 case, the current design used today for the default primary network will need some changes. Since Egress IP may be served on multiple namespaces and thus networks, it is possible that there could be a collision as previously mentioned in the Pod Egress section. Therefore, the same solution provided in that section where the GR SNATs to the masquerade subnet must be utilized. However, once the packet arrives in br-ex we will need a way to tell if it was sent from a pod affected by a specific egress IP. To address this, pkt_mark will be used to mark egress IP packets and signify to br-ex which egress IP to SNAT to. An example where the egress IP is 1.1.1.1 that maps to pkt_mark 10 would look something like this:
```

Comment:
@trozet keep in mind that we do **not** support secondary L3 in openshift. There's no mention of it anywhere (docs), and I even think it wasn't QE'd.

I'm not saying we shouldn't, I'm simply stating that it'll imply more work (for docs / QE teams, since the code is already available ...).

Reply:
ok upstream I think it makes sense to support it, as there should be no difference to use really if it is for primary or secondary use. For downstream support, we can make that call later.